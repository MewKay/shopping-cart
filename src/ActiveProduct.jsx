import PropTypes from "prop-types";
import { useState } from "react";
import { useOutletContext } from "react-router-dom";

const ActiveProduct = ({ product, onRemoveActiveProduct }) => {
  const { cart, setCart } = useOutletContext();
  const [productQuantity, setProductQuantity] = useState(1);
  const { id, title, image, description, price } = product;

  const handleIncreaseQuantity = () =>
    !productQuantity
      ? setProductQuantity(1)
      : setProductQuantity((value) => Number.parseInt(value) + 1);

  const handleDecreaseQuantity = () =>
    productQuantity <= 1 || !productQuantity
      ? setProductQuantity(1)
      : setProductQuantity((value) => Number.parseInt(value) - 1);

  const handleTypeQuantity = (e) => {
    const typedValue = e.target.value;
    setProductQuantity(typedValue);
  };

  const handleAddItemToCart = () => {
    const newCart = structuredClone(cart);
    const indexItemToAdd = newCart.findIndex(
      (cartItem) => cartItem.productDetails.id === id
    );

    if (indexItemToAdd >= 0) {
      newCart[indexItemToAdd].quantity += productQuantity;
    } else {
      newCart.push({
        productDetails: product,
        quantity: Number.parseInt(productQuantity),
      });
    }

    setCart(newCart);
  };

  const handleCloseSelectedItem = () => {
    onRemoveActiveProduct();
    setProductQuantity(1);
  };

  return (
    <div>
      <div>
        <img src={image} alt="" />
      </div>
      <div>
        <h3>{title}</h3>
        <p>{description}</p>
        <p>${price}</p>
        <div>
          <button
            aria-label="Increase quantity"
            onClick={handleIncreaseQuantity}
          >
            +
          </button>
          <input
            aria-label="Quantity value"
            type="number"
            min={1}
            value={productQuantity}
            onChange={handleTypeQuantity}
          />
          <button
            aria-label="Decrease quantity"
            onClick={handleDecreaseQuantity}
          >
            -
          </button>
        </div>
      </div>
      <div>
        <button onClick={handleAddItemToCart}>Add to cart</button>
      </div>
      <button
        aria-label="Close selected product"
        onClick={handleCloseSelectedItem}
      >
        x
      </button>
    </div>
  );
};

ActiveProduct.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    image: PropTypes.string,
    description: PropTypes.string,
    price: PropTypes.number,
  }).isRequired,
  onRemoveActiveProduct: PropTypes.func.isRequired,
};

export default ActiveProduct;
