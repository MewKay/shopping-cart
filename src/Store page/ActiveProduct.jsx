import PropTypes from "prop-types";
import { useState } from "react";
import { useOutletContext } from "react-router-dom";
import { CircleX, Minus, Plus, ShoppingCart } from "lucide-react";
import styles from "./ActiveProduct.module.css";

const ActiveProduct = ({ product, onRemoveActiveProduct }) => {
  const { cart, setCart } = useOutletContext();
  const [productQuantity, setProductQuantity] = useState(1);
  const { id, title, image, description, price } = product;

  const handleIncreaseQuantity = () =>
    !productQuantity
      ? setProductQuantity(1)
      : setProductQuantity((value) => Number.parseInt(value) + 1);

  const handleDecreaseQuantity = () =>
    productQuantity <= 1 || !productQuantity
      ? setProductQuantity(1)
      : setProductQuantity((value) => Number.parseInt(value) - 1);

  const handleTypeQuantity = (e) => {
    const typedValue = e.target.value;
    setProductQuantity(typedValue);
  };

  const handleAddItemToCart = () => {
    const newCart = structuredClone(cart);
    const indexItemToAdd = newCart.findIndex(
      (cartItem) => cartItem.productDetails.id === id
    );

    if (indexItemToAdd >= 0) {
      newCart[indexItemToAdd].quantity += productQuantity;
    } else {
      newCart.push({
        productDetails: product,
        quantity: Number.parseInt(productQuantity),
      });
    }

    setCart(newCart);
  };

  const handleCloseSelectedItem = () => {
    onRemoveActiveProduct();
    setProductQuantity(1);
  };

  return (
    <div className={styles["container"]}>
      <div className={styles["image-container"]}>
        <img src={image} alt="" className={styles["product-preview"]} />
      </div>
      <div className={styles["product-info"]}>
        <h3 className={styles["title"]}>{title}</h3>
        <p className={styles["description"]}>{description}</p>
        <p className={styles["price"]}>${price}</p>
        <div className={styles["quantity-container"]}>
          <button
            aria-label="Increase quantity"
            onClick={handleIncreaseQuantity}
            className={styles["add-button"]}
          >
            <Plus />
          </button>
          <input
            aria-label="Quantity value"
            type="number"
            min={1}
            max={999}
            value={productQuantity}
            onChange={handleTypeQuantity}
            className={styles["quantity-input"]}
          />
          <button
            aria-label="Decrease quantity"
            onClick={handleDecreaseQuantity}
            className={styles["subtract-button"]}
          >
            <Minus />
          </button>
        </div>
        <div className={styles["add-cart-container"]}>
          <button
            onClick={handleAddItemToCart}
            className={styles["add-cart-button"]}
          >
            Add to cart
            <ShoppingCart />
          </button>
        </div>
      </div>
      <button
        aria-label="Close selected product"
        onClick={handleCloseSelectedItem}
        className={styles["close-button"]}
      >
        <CircleX />
      </button>
    </div>
  );
};

ActiveProduct.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    image: PropTypes.string,
    description: PropTypes.string,
    price: PropTypes.number,
  }).isRequired,
  onRemoveActiveProduct: PropTypes.func.isRequired,
};

export default ActiveProduct;
