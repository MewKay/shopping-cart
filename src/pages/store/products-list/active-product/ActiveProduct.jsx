import PropTypes from "prop-types";
import { useState } from "react";
import { useOutletContext } from "react-router-dom";
import { CircleX, ShoppingCart } from "lucide-react";
import { motion } from "motion/react";
import styles from "./ActiveProduct.module.css";
import QuantitySelector from "../../../../components/quantity-selector/QuantitySelector";

const ActiveProduct = ({ product, onRemoveActiveProduct }) => {
  const { cart, setCart } = useOutletContext();
  const [productQuantity, setProductQuantity] = useState(1);
  const { id, title, image, description, price } = product;

  const handleIncreaseQuantity = () =>
    !productQuantity
      ? setProductQuantity(1)
      : setProductQuantity((value) => Number.parseInt(value) + 1);

  const handleDecreaseQuantity = () =>
    productQuantity <= 1 || !productQuantity
      ? setProductQuantity(1)
      : setProductQuantity((value) => Number.parseInt(value) - 1);

  const handleTypeQuantity = (e) => {
    const typedValue = e.target.value;
    setProductQuantity(typedValue);
  };

  const handleAddItemToCart = () => {
    const newCart = structuredClone(cart);
    const indexItemToAdd = newCart.findIndex(
      (cartItem) => cartItem.productDetails.id === id
    );

    if (indexItemToAdd >= 0) {
      newCart[indexItemToAdd].quantity += Number.parseInt(productQuantity);
    } else {
      newCart.push({
        productDetails: product,
        quantity: Number.parseInt(productQuantity),
      });
    }

    setCart(newCart);
  };

  const handleCloseSelectedItem = () => {
    onRemoveActiveProduct();
    setProductQuantity(1);
  };

  return (
    <motion.div
      initial={{ clipPath: "inset(0 0 100% 0)", height: 0 }}
      animate={{ clipPath: "inset(0)", height: "auto" }}
      exit={{ clipPath: "inset(0 0 100% 0)", height: 0 }}
      transition={{ type: "tween", duration: 0.3 }}
      className={styles["animation-wrapper"]}
    >
      <div className={styles["container"]}>
        <div className={styles["image-container"]}>
          <img src={image} alt="" className={styles["product-preview"]} />
        </div>
        <div className={styles["product-info"]}>
          <h3 className={styles["title"]}>{title}</h3>
          <p className={styles["description"]}>{description}</p>
          <p className={styles["price"]}>${price}</p>
          <QuantitySelector
            productQuantity={productQuantity}
            onIncreaseQuantity={handleIncreaseQuantity}
            onTypeQuantity={handleTypeQuantity}
            onDecreaseQuantity={handleDecreaseQuantity}
          />
          <div className={styles["add-cart-container"]}>
            <button
              onClick={handleAddItemToCart}
              className={styles["add-cart-button"]}
            >
              Add to cart
              <ShoppingCart />
            </button>
          </div>
        </div>
        <button
          aria-label="Close selected product"
          onClick={handleCloseSelectedItem}
          className={styles["close-button"]}
        >
          <CircleX />
        </button>
      </div>
    </motion.div>
  );
};

ActiveProduct.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.number,
    title: PropTypes.string,
    image: PropTypes.string,
    description: PropTypes.string,
    price: PropTypes.number,
  }).isRequired,
  onRemoveActiveProduct: PropTypes.func.isRequired,
};

export default ActiveProduct;
