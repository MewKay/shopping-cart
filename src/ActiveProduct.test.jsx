import { describe, it, expect, vi } from "vitest";
import { render, screen } from "@testing-library/react";
import ActiveProduct from "./ActiveProduct";
import userEvent from "@testing-library/user-event";

//Component props
const product = {
  id: 0,
  title: "Product 0",
  image: "https://image.placeholder.com/product",
  description:
    "Lorem ipsum dolor sit amet consectetur adipisicing elit. Delectus magnam asperiores minus dolores cumque corporis vel dolorem atque non reiciendis.",
  price: 50,
};
const onAddToCart = vi.fn();
const onRemoveActiveProduct = vi.fn();

describe("ActiveProduct component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <ActiveProduct
        product={product}
        onAddToCart={onAddToCart}
        onRemoveActiveProduct={onRemoveActiveProduct}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("calls 'onAddToCart' when the Add to cart button is clicked", async () => {
    const user = userEvent.setup();
    render(
      <ActiveProduct
        product={product}
        onAddToCart={onAddToCart}
        onRemoveActiveProduct={onRemoveActiveProduct}
      />
    );

    const addToCartButton = screen.getByRole("button", { name: "Add to cart" });
    await user.click(addToCartButton);

    expect(onAddToCart).toHaveBeenCalled();
  });

  it("calls 'onRemoveActiveProduct' when the close button is clicked", async () => {
    const user = userEvent.setup();
    render(
      <ActiveProduct
        product={product}
        onAddToCart={onAddToCart}
        onRemoveActiveProduct={onRemoveActiveProduct}
      />
    );

    const closeActiveProductButton = screen.getByLabelText(
      "Close selected product"
    );
    await user.click(closeActiveProductButton);

    expect(onRemoveActiveProduct).toHaveBeenCalled();
  });

  describe("Product quantity", () => {
    it("should have 0 as a default value", async () => {
      render(
        <ActiveProduct
          product={product}
          onAddToCart={onAddToCart}
          onRemoveActiveProduct={onRemoveActiveProduct}
        />
      );

      const quantityInput = screen.getByLabelText("Quantity value");

      expect(quantityInput).toHaveValue(0);
    });

    it("increase the value by one when 'Increase quantity' button is clicked", async () => {
      const user = userEvent.setup();
      render(
        <ActiveProduct
          product={product}
          onAddToCart={onAddToCart}
          onRemoveActiveProduct={onRemoveActiveProduct}
        />
      );

      const quantityInput = screen.getByLabelText("Quantity value");
      const quantityInputInitialValue = Number.parseInt(quantityInput.value);
      const increaseQuantityButton = screen.getByLabelText("Increase quantity");

      await user.click(increaseQuantityButton);

      expect(quantityInput).toHaveValue(quantityInputInitialValue + 1);
    });

    it("decrease the value by one when 'Decrease quantity' button is clicked", async () => {
      const user = userEvent.setup();
      render(
        <ActiveProduct
          product={product}
          onAddToCart={onAddToCart}
          onRemoveActiveProduct={onRemoveActiveProduct}
        />
      );

      const quantityInput = screen.getByLabelText("Quantity value");
      await user.type(quantityInput, "10");
      const quantityInputInitialValue = Number.parseInt(quantityInput.value);
      const decreaseQuantityButton = screen.getByLabelText("Decrease quantity");

      await user.click(decreaseQuantityButton);

      expect(quantityInput).toHaveValue(quantityInputInitialValue - 1);
    });
  });
});
