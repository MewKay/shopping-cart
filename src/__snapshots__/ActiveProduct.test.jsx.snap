// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ActiveProduct component > renders correctly 1`] = `
<div>
  <div>
    <div>
      <img
        alt=""
        src="https://image.placeholder.com/product"
      />
    </div>
    <div>
      <h3>
        Product 0
      </h3>
      <p>
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Delectus magnam asperiores minus dolores cumque corporis vel dolorem atque non reiciendis.
      </p>
      <p>
        $
        50
      </p>
      <div>
        <button
          aria-label="Increase quantity"
        >
          +
        </button>
        <input
          aria-label="Quantity value"
          min="1"
          type="number"
          value="1"
        />
        <button
          aria-label="Decrease quantity"
        >
          -
        </button>
      </div>
    </div>
    <div>
      <button>
        Add to cart
      </button>
    </div>
    <button
      aria-label="Close selected product"
    >
      x
    </button>
  </div>
</div>
`;
